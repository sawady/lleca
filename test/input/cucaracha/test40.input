

fun esCero(n : Int) : Bool {
    return n == 0
}

fun producto(x : Int, y : Int) : Int {
    return x * y
}

fun factorial(n : Int) : Int {
    if esCero(n) {
        n := 1
    } else {
        n := n - 1
        n := producto(n + 1, factorial(n))
    }
    return n
}

fun main() {
    putNum(factorial(1)) putChar(10)
    putNum(factorial(2)) putChar(10)
    putNum(factorial(3)) putChar(10)
    putNum(factorial(4)) putChar(10)
    putNum(factorial(5)) putChar(10)
    putNum(factorial(6)) putChar(10)
    putNum(factorial(7)) putChar(10)
    putNum(factorial(8)) putChar(10)
    putNum(factorial(9)) putChar(10)
    putNum(factorial(10)) putChar(10)
}
