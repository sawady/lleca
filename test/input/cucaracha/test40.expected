Cons(
  Function(
    esCero,
    Cons(
      Parameter(
        n,
        Int
      ),
      Nil
    ),
    Bool,
    Cons(
      StmtReturn(
        ExprEq(
          ExprVar(
            n
          ),
          ExprConstNum(
            0
          )
        )
      ),
      Nil
    )
  ),
  Cons(
    Function(
      producto,
      Cons(
        Parameter(
          x,
          Int
        ),
        Cons(
          Parameter(
            y,
            Int
          ),
          Nil
        )
      ),
      Int,
      Cons(
        StmtReturn(
          ExprMul(
            ExprVar(
              x
            ),
            ExprVar(
              y
            )
          )
        ),
        Nil
      )
    ),
    Cons(
      Function(
        factorial,
        Cons(
          Parameter(
            n,
            Int
          ),
          Nil
        ),
        Int,
        Cons(
          StmtIf(
            ExprCall(
              esCero,
              Cons(
                ExprVar(
                  n
                ),
                Nil
              )
            ),
            Cons(
              StmtAssign(
                n,
                ExprConstNum(
                  1
                )
              ),
              Nil
            ),
            Just(
              Cons(
                StmtAssign(
                  n,
                  ExprSub(
                    ExprVar(
                      n
                    ),
                    ExprConstNum(
                      1
                    )
                  )
                ),
                Cons(
                  StmtAssign(
                    n,
                    ExprCall(
                      producto,
                      Cons(
                        ExprAdd(
                          ExprVar(
                            n
                          ),
                          ExprConstNum(
                            1
                          )
                        ),
                        Cons(
                          ExprCall(
                            factorial,
                            Cons(
                              ExprVar(
                                n
                              ),
                              Nil
                            )
                          ),
                          Nil
                        )
                      )
                    )
                  ),
                  Nil
                )
              )
            )
          ),
          Cons(
            StmtReturn(
              ExprVar(
                n
              )
            ),
            Nil
          )
        )
      ),
      Cons(
        Function(
          main,
          Nil,
          Unit,
          Cons(
            StmtCall(
              putNum,
              Cons(
                ExprCall(
                  factorial,
                  Cons(
                    ExprConstNum(
                      1
                    ),
                    Nil
                  )
                ),
                Nil
              )
            ),
            Cons(
              StmtCall(
                putChar,
                Cons(
                  ExprConstNum(
                    10
                  ),
                  Nil
                )
              ),
              Cons(
                StmtCall(
                  putNum,
                  Cons(
                    ExprCall(
                      factorial,
                      Cons(
                        ExprConstNum(
                          2
                        ),
                        Nil
                      )
                    ),
                    Nil
                  )
                ),
                Cons(
                  StmtCall(
                    putChar,
                    Cons(
                      ExprConstNum(
                        10
                      ),
                      Nil
                    )
                  ),
                  Cons(
                    StmtCall(
                      putNum,
                      Cons(
                        ExprCall(
                          factorial,
                          Cons(
                            ExprConstNum(
                              3
                            ),
                            Nil
                          )
                        ),
                        Nil
                      )
                    ),
                    Cons(
                      StmtCall(
                        putChar,
                        Cons(
                          ExprConstNum(
                            10
                          ),
                          Nil
                        )
                      ),
                      Cons(
                        StmtCall(
                          putNum,
                          Cons(
                            ExprCall(
                              factorial,
                              Cons(
                                ExprConstNum(
                                  4
                                ),
                                Nil
                              )
                            ),
                            Nil
                          )
                        ),
                        Cons(
                          StmtCall(
                            putChar,
                            Cons(
                              ExprConstNum(
                                10
                              ),
                              Nil
                            )
                          ),
                          Cons(
                            StmtCall(
                              putNum,
                              Cons(
                                ExprCall(
                                  factorial,
                                  Cons(
                                    ExprConstNum(
                                      5
                                    ),
                                    Nil
                                  )
                                ),
                                Nil
                              )
                            ),
                            Cons(
                              StmtCall(
                                putChar,
                                Cons(
                                  ExprConstNum(
                                    10
                                  ),
                                  Nil
                                )
                              ),
                              Cons(
                                StmtCall(
                                  putNum,
                                  Cons(
                                    ExprCall(
                                      factorial,
                                      Cons(
                                        ExprConstNum(
                                          6
                                        ),
                                        Nil
                                      )
                                    ),
                                    Nil
                                  )
                                ),
                                Cons(
                                  StmtCall(
                                    putChar,
                                    Cons(
                                      ExprConstNum(
                                        10
                                      ),
                                      Nil
                                    )
                                  ),
                                  Cons(
                                    StmtCall(
                                      putNum,
                                      Cons(
                                        ExprCall(
                                          factorial,
                                          Cons(
                                            ExprConstNum(
                                              7
                                            ),
                                            Nil
                                          )
                                        ),
                                        Nil
                                      )
                                    ),
                                    Cons(
                                      StmtCall(
                                        putChar,
                                        Cons(
                                          ExprConstNum(
                                            10
                                          ),
                                          Nil
                                        )
                                      ),
                                      Cons(
                                        StmtCall(
                                          putNum,
                                          Cons(
                                            ExprCall(
                                              factorial,
                                              Cons(
                                                ExprConstNum(
                                                  8
                                                ),
                                                Nil
                                              )
                                            ),
                                            Nil
                                          )
                                        ),
                                        Cons(
                                          StmtCall(
                                            putChar,
                                            Cons(
                                              ExprConstNum(
                                                10
                                              ),
                                              Nil
                                            )
                                          ),
                                          Cons(
                                            StmtCall(
                                              putNum,
                                              Cons(
                                                ExprCall(
                                                  factorial,
                                                  Cons(
                                                    ExprConstNum(
                                                      9
                                                    ),
                                                    Nil
                                                  )
                                                ),
                                                Nil
                                              )
                                            ),
                                            Cons(
                                              StmtCall(
                                                putChar,
                                                Cons(
                                                  ExprConstNum(
                                                    10
                                                  ),
                                                  Nil
                                                )
                                              ),
                                              Cons(
                                                StmtCall(
                                                  putNum,
                                                  Cons(
                                                    ExprCall(
                                                      factorial,
                                                      Cons(
                                                        ExprConstNum(
                                                          10
                                                        ),
                                                        Nil
                                                      )
                                                    ),
                                                    Nil
                                                  )
                                                ),
                                                Cons(
                                                  StmtCall(
                                                    putChar,
                                                    Cons(
                                                      ExprConstNum(
                                                        10
                                                      ),
                                                      Nil
                                                    )
                                                  ),
                                                  Nil
                                                )
                                              )
                                            )
                                          )
                                        )
                                      )
                                    )
                                  )
                                )
                              )
                            )
                          )
                        )
                      )
                    )
                  )
                )
              )
            )
          )
        ),
        Nil
      )
    )
  )
)
